lib/spack/docs/build_systems/intelpackage.rst

```diff
@@ -0,0 +1,994 @@
+             operating_system:   centos6
+             modules:    []
+             spec:       intel@18.0.2
+             paths:
+               cc:       stub
+               cxx:      stub
+               f77:      stub
+               fc:       stub
+
+      Replace ``18.0.2`` with the version that you determined in the preceeding
+      step. The contents of the language compiler tags (``cc:`` etc.) do not
+      matter at this point.
+
+      **Note:** If you already have a certain ``%intel@x.y.z`` compiler spec in
+      place and you wish to re-install the ``intel-parallel-studio`` (or
+      ``intel``) package providing the *same* compiler version, you do not need
+      to revert its ``compilers.yaml`` declaration to stub form as shown here.
+      When done, however, you may still need to adjust the entries under the
+      ``paths:`` tag (`see below <Post-install steps for compilers_>`_) if the
+      package's installation directory changed, such as in the hash portion.
+
+   .. _`verify-compiler-anticipated`:
+
+   C. Verify that the new compiler version will be used as expected:
+
+      You should see it if you placed the stub last in ``compilers.yaml`` and
+      ask for the compiler just by name, e.g.:
+
+      .. code-block:: sh
+
+         spack spec zlib %intel
+
+      Otherwise, or simply to be explicit, state the anticipated compiler
+      version as well, e.g.:
+
+      .. code-block:: sh
+
+         spack spec zlib %intel@18.0.2
+
+   You are right to ask: "Why on earth is that necessary?" [fn9]_.
+   The answer lies in Spack striving for strict compiler consistency.
+   Consider what happens without a pre-declared compiler stub:
+   You ask Spack to install a particular version
+   ``intel-parallel-studio@edition.V``.  Spack will apply an unrelated compiler
+   spec to concretize your request, giving ``intel-parallel-studio@edition.V
+   %X``. Naturally, ``%X`` is not going to be the version that this new package
+   provides, but typically ``%gcc@...`` in a default Spack installation or possibly
+   indeed ``%intel@...``, though at a version preceeding ``V``.
+
+   The problem comes to the fore as soon as you try to use any virtual
+   ``mkl`` or ``mpi`` packages that you would expect to now be provided by
+   ``intel-parallel-studio@edition.V``.  Spack will indeed see those virtual packages,
+   but only as being tied to the compiler concretized *at installation*.
+   If you were to install a client package with the new compilers now
+   available to you, you would run ``spack install foo +mkl %intel@V``, but
+   Spack would complain about ``mkl%intel@V`` being missing, because it only
+   knows about ``mkl%X``.
+
+   To escape this trap, put the compiler stub declaration shown here in place,
+   then use that pre-declared compiler spec to install the actual package, as
+   shown in the next step.  This approach works because only the package's
+   builtin binary installer will be used, not any of the compilers.
```

Comment:
> You really should be installing all compilers with the system "core compiler" if you want to use LMod

Why? Or do you mean what you mark as the core compiler in `modules.yaml`?

I didn't use the default system compiler when I set up a RHEL 6 cluster for LMod, it was _ancient_. I used one from one of the devtoolset rpms and it worked fine.

Reply:
Yeah, whatever the "core compiler" in `modules.yaml` is. If you can get away with it, I prefer to use the system compiler because I always see problems with Intel pulling in GCC libs that it isn't supposed to.

Reply:
@adamjstewart 
> Are you suggesting that people create a fake compiler of the same version they are trying to install and then install that version with that compiler?

Yes, that's absolutely what I suggest here.

As you mention, it does solve the re-install issues, but I didn't know about how LMod is affected. What's considered the "core compiler" for LMod, and what is the "system compiler" as far as Spack is concerned? If I have both `gcc` and `icc` in `PATH` and do `spack compilers` the first time, does `gcc` get chosen as default merely because of ASCII order?

Reply:
When I say "system compiler", I mean the default compiler that comes with the operating system after a clean installation, usually GCC, or Clang on macOS. This isn't really a Spack thing, I just made up the term. It is often the same as the LMod "core compiler", but as @svenevs pointed out, it doesn't have to be.

When I say "core compiler", this is an LMod term. From LMod's documentation, they describe the [Core/Compiler/MPI hierarchy](http://lmod.readthedocs.io/en/latest/080_hierarchy.html) as:

> The Core directory is for modules that are not dependent on Compiler or MPI implementations. The Compiler directory is for packages which are only Compiler dependent. Lastly, the MPI directory is packages which dependent on MPI-Compiler pairing. The modulefiles for the compilers are placed in the Core directory.

When you start using LMod with Spack, you have to explicitly choose which of the compilers defined in `compilers.yaml` is the "core compiler", Spack doesn't choose for you. My usual setup, and what I think LMod intends, is that you take your ancient version of GCC on the system and consider that to be the "core compiler". The only things you should install with this compiler are things that are "relatively" compiler independent. For example, builds tools like CMake and Autotools and newer compilers like GCC or Intel. You then add those compilers to Spack and proceed to build the things you actually need with them. This might include MPI. You then build things with that MPI installation. 

Spack automatically generates the Core/Compiler/MPI module file hierarchy for you as long as your "core compiler" is defined in `modules.yaml`. When you first log into a system, unless a default compiler/MPI is chosen for you, `module avail` only lists compilers (and maybe build tools). Once you `module load` a compiler, it adds the directory of modules built with that compiler to your `MODULEPATH`. Once you `module load` an MPI provider, it prepends another directory with all the modules built with that MPI. This beautiful, chaotic system is designed to prevent you from loading incompatible packages, like an HDF5 library built with one compiler and MPI implementation, and a NetCDF library built with another.

> If I have both `gcc` and `icc` in `PATH` and do `spack compilers` the first time, does `gcc` get chosen as default merely because of ASCII order?

Nope, the default order is explicitly defined in [`packages.yaml`](https://github.com/spack/spack/blob/develop/etc/spack/defaults/packages.yaml#L18). Obviously, you can override this. By default, Spack uses the newest GCC compiler, although once upon a time it used the first GCC compiler listed in `compilers.yaml`. I'm pretty sure that bug was fixed though.